### PR TITLE
Refactor Celestial Anchors

### PR DESCRIPTION
**Pull request by Schneegans**
This major refactoring reduces the CPU load of scenes containing many celestial bodies significantly. The primary change is a shift from inheritance to composition when it comes to the celestial anchors. For the default scene, the time to update the `SolarSystem` has been reduced by more than 50% (on my machine from 11 ms to 5 ms).

# Primary Changes

* The `Settings` class now has a map of immutable `CelestialObject`s which are instantiated when CosmoScout is started. Before it used to have a list of anchor configurations which could be used to initialize `CelestialObject`s.
* `CelestialObject`s now have four new properties:
  * `bodyCullingRadius`: This can be used to determine the visiblity of an object. If the object is too far away to be seen, plugins may want to skip updating or drawing attached things.
  * `orbitCullingRadius`:This can be used to determine the visiblity of an object's trajectory. If the object is really far away, plugins may want to skip updating or drawing attached things (like trajectories or anchor labels).
  * `trackable`: If set to `false`, the observer will not automatically follow this object.
  * `collidable`:  If set to `false`, the observer can travel through the object. 
* The observer-relative transformation of each `CelestialObject` is updated once each frame by the `SolarSystem`.
* All classes which previously derived from `CelestialAnchor`, `CelestialObject`, or `CelestialBody` now get the observer-relative transformation from the `CelestialObject`s in the `Settings`.
* Classes which previously derived from `CelestialBody` now have to derive from the new class `CelestialSurface`. A `CelestialSurface` can then be assigned to a `CelestialObject`.
* `CelestialAnchorNode`s do not exist anymore. Classes which  used to use these, now have to use a `VistaTransformNode` instead and update its transformation once each frame with the observer-relative transformation from the respective  `CelestialObject`.
* Much per-frame logic used to be executed in overrides of `CelestialObject::update()`. This method does not exist any more; plugins now need to update their objects manually.
* The Windows CI jobs do not use `clcache` anymore. See #299.

# Other Changes

* The mouse navigation has been made a bit smoother. This makes it possible to update the scene scale only once each frame. Before, it had to be updated multiple times each frame which introduced a significant CPU overhead during rapid movements.
* All plugins now save their configuration before being unloaded. This makes it possible that plugins keep their current state when being reloaded at run-time.
* Test source files are now only compiled into the executable and into the libraries if `COSMOSCOUT_UNIT_TESTS` is set to `true`.

### `csp-anchor-labels`

* Anchor Labels are now shown for each configured object per default. A blacklist has been added to the plugin configuration to remove labels selectively.
* Anchor Labels now show the name of the configured object rather than the SPICE center name.
* The maximum size of Anchor Labels has been increased to prevent line breaking for longer object names.
* Anchor Labels now use the new `orbitCullingRadius` of the object for deciding whether they should be shown or not.

### `csp-measurement-tools`

* Some common code has been moved from the measurement tools to the base classes `cs::core::tools::Tool` and `cs::core::tools::Mark`. This simplifies some code in the measurement tools.

### `csp-satellites`

* It's now possible to reload this plugin at run-time.

### `csp-wms-overlays`

* A regression was fixed which led to low precision close to a bodies surface.

# Future Work

* The method `SolarSystem::getObject()` is called pretty frequently. Performance-wise it could be better to store `weak_ptr`s to the  `CelestialObject`s instead of calling this method every frame. For now, most objects only store the name of the  `CelestialObject` they are attached to and get the corresponding `CelestialObject` from the settings each frame.
* If a plugin references an object which is actually not available (e.g. due to a typo in the scene configuration), CosmoScout will crash (it prints a error message though).
* Adding and removing `CelestialObject`s at run-time could work now (like adding a completely new planet at run-time), but has not been tested. Plugins may need some tweaking to properly respond to this.

### CODE REVIEW

File: config/base/scene/simple_desktop.json [hunk 1]
```diff
@@ -810,18 +882,6 @@
...
-      "preFetch": 1,
-      "bodies": {
-        "Earth": {
-          "wms": [
-            "https://maps.dwd.de/geoserver/dwd/wms",
-            "https://neo.sci.gsfc.nasa.gov/wms/wms"
-          ]
-        }
-      }
-    },
```
Comment (JonasGilg) [inline config/base/scene/simple_desktop.json:824]: Why was this removed?

Comment (Schneegans) [reply config/base/scene/simple_desktop.json:824]: Mostly, because both servers did not work at the time I tested it. Maybe we should test it again. But a larger wall of errors does not bring a good user experience for the default configuration 😉

**Comment by JonasGilg** [comment]
The csp-satellites plugin throws the following OpenGL errors:

```
[I] csp-satellites .... Loading plugin...
[W] cs-graphics ....... From vista-gltf: OpenGL error in "createGPUsampler": 1280 (1280) C:/C++/cs-gh/src/cs-graphics/internal/gltfmodel.cpp:916
[W] cs-graphics ....... From vista-gltf: OpenGL error in "createGPUsampler": 1280 (1280) C:/C++/cs-gh/src/cs-graphics/internal/gltfmodel.cpp:916
[W] cs-graphics ....... From vista-gltf: OpenGL error in "createGPUsampler": 1280 (1280) C:/C++/cs-gh/src/cs-graphics/internal/gltfmodel.cpp:916
[I] csp-satellites .... Loading done.
```

**Comment by Schneegans** [comment]
> The csp-satellites plugin throws the following OpenGL errors:
> 
> ```
> [I] csp-satellites .... Loading plugin...
> [W] cs-graphics ....... From vista-gltf: OpenGL error in "createGPUsampler": 1280 (1280) C:/C++/cs-gh/src/cs-graphics/internal/gltfmodel.cpp:916
> [W] cs-graphics ....... From vista-gltf: OpenGL error in "createGPUsampler": 1280 (1280) C:/C++/cs-gh/src/cs-graphics/internal/gltfmodel.cpp:916
> [W] cs-graphics ....... From vista-gltf: OpenGL error in "createGPUsampler": 1280 (1280) C:/C++/cs-gh/src/cs-graphics/internal/gltfmodel.cpp:916
> [I] csp-satellites .... Loading done.
> ```

This is true, but it should be a separate issue, as it also happens on the current `develop` branch.

File: src/cs-core/EclipseShadowReceiver.cpp [hunk 2]
```diff
@@ -78,17 +76,16 @@ void EclipseShadowReceiver::update(double time, scene::CelestialObserver const&
   }
 
   // Acquire a list of allpotentially relevant eclipse shadow maps.
```
Comment (JonasGilg) [inline src/cs-core/EclipseShadowReceiver.cpp:78]:
```suggestion
  // Acquire a list of all potentially relevant eclipse shadow maps.
```

File: src/cs-core/tools/DeletableMark.cpp [hunk 3]
```diff
@@ -24,9 +25,8 @@ namespace cs::core::tools {
 
 DeletableMark::DeletableMark(std::shared_ptr<InputManager> const& pInputManager,
     std::shared_ptr<SolarSystem> const& pSolarSystem, std::shared_ptr<Settings> const& settings,
-    std::shared_ptr<TimeControl> const& pTimeControl, std::string const& sCenter,
-    std::string const& sFrame)
-    : Mark(pInputManager, pSolarSystem, settings, pTimeControl, sCenter, sFrame)
+    std::string const& objectName)
+    : Mark(pInputManager, pSolarSystem, settings, objectName)
```
Comment (JonasGilg) [inline src/cs-core/tools/DeletableMark.cpp:29]: Pass by value and use `std::move`.

File: src/cs-core/tools/Mark.cpp [hunk 4]
```diff
@@ -65,18 +67,18 @@ void main()
 ////////////////////////////////////////////////////////////////////////////////////////////////////
 
 Mark::Mark(std::shared_ptr<InputManager> pInputManager, std::shared_ptr<SolarSystem> pSolarSystem,
-    std::shared_ptr<Settings> settings, std::shared_ptr<TimeControl> pTimeControl,
-    std::string const& sCenter, std::string const& sFrame)
-    : mInputManager(std::move(pInputManager))
+    std::shared_ptr<Settings> settings, std::string const& objectName)
+    : Tool(objectName)
```
Comment (JonasGilg) [inline src/cs-core/tools/Mark.cpp:71]:
```suggestion
    std::shared_ptr<Settings> settings, std::string const& objectName)
    : Tool(objectName)
```
```suggestion
    std::shared_ptr<Settings> settings, std::string objectName)
    : Tool(std::move(objectName))
```

File: src/cs-core/tools/MultiPointTool.cpp [hunk 5]
```diff
@@ -19,13 +19,11 @@ namespace cs::core::tools {
 
 MultiPointTool::MultiPointTool(std::shared_ptr<InputManager> pInputManager,
     std::shared_ptr<SolarSystem> pSolarSystem, std::shared_ptr<Settings> settings,
-    std::shared_ptr<TimeControl> pTimeControl, std::string sCenter, std::string sFrame)
-    : mInputManager(std::move(pInputManager))
+    std::string objectName)
+    : Tool(objectName)
```
Comment (JonasGilg) [inline src/cs-core/tools/MultiPointTool.cpp:23]:
```suggestion
    : Tool(std::move(objectName))
```

File: src/cs-core/tools/MultiPointTool.cpp [hunk 6]
```diff
@@ -123,51 +117,35 @@ void MultiPointTool::update() {
...
-std::string const& MultiPointTool::getCenterName() const {
-  return mCenter;
-}
-
-////////////////////////////////////////////////////////////////////////////////////////////////////
-
-void MultiPointTool::setFrameName(std::string const& name) {
-  mFrame = name;
+void MultiPointTool::setObjectName(std::string const& name) {
+  Tool::setObjectName(name);
```
Comment (JonasGilg) [inline src/cs-core/tools/MultiPointTool.cpp:140]:
```suggestion
void MultiPointTool::setObjectName(std::string name) {
  Tool::setObjectName(std::move(name));
```

File: src/cs-core/tools/MultiPointTool.cpp [hunk 7]
```diff
@@ -123,51 +117,35 @@ void MultiPointTool::update() {
...
 
-  for (auto& p : mPoints) {
-    p->getAnchor()->setFrameName(name);
+  for (auto point : mPoints) {
+    point->setObjectName(name);
```
Comment (JonasGilg) [inline src/cs-core/tools/MultiPointTool.cpp:143]:
```suggestion
  for (auto const& point : mPoints) {
    point->setObjectName(getObjectName());
```

File: src/cs-core/tools/MultiPointTool.hpp [hunk 8]
```diff
@@ -55,13 +55,8 @@ class CS_CORE_EXPORT MultiPointTool : public Tool {
...
 
-  /// Gets or sets the SPICE center name for all points.
-  virtual void               setCenterName(std::string const& name);
-  virtual std::string const& getCenterName() const;
-
-  /// Gets or sets the SPICE frame name for all points.
-  virtual void               setFrameName(std::string const& name);
-  virtual std::string const& getFrameName() const;
+  // Assigns all points to a new celestial object.
+  void setObjectName(std::string const& name) override;
```
Comment (JonasGilg) [inline src/cs-core/tools/MultiPointTool.hpp:59]:
```suggestion
  void setObjectName(std::string name) override;
```

**Review by JonasGilg — CHANGES_REQUESTED** [review]
Amazing work! The performance improvements are incredible. I couldn't find any major problems in the changes, I just have some minor suggestions for improvements.

File: plugins/csp-custom-web-ui/src/Plugin.cpp [hunk 9]
```diff
@@ -133,21 +128,29 @@ void Plugin::init() {
 ////////////////////////////////////////////////////////////////////////////////////////////////////
 
 void Plugin::update() {
-  double simulationTime(mTimeControl->pSimulationTime.get());
 
   // Rotate the space items to face the observer.
   for (auto& item : mSpaceItems) {
```
[the remark this answers is not in the document]
Comment (Schneegans) [reply plugins/csp-custom-web-ui/src/Plugin.cpp:133]: What would be the benefit here? Shouldn't be this the same in the context?

File: plugins/csp-measurement-tools/src/PathTool.cpp [hunk 10]
```diff
@@ -233,36 +207,16 @@ void PathTool::updateLineVertices() {
...
-    } catch (std::exception const& e) {
-      // Getting the relative transformation may fail due to insufficient SPICE data.
-      logger().warn("Failed to calculate scale distance of Path Tool: {}", e.what());
-    }
+    mPosition += mark->getPosition() / static_cast<double>(mPoints.size());
   }
 
-  auto lastMark = mPoints.begin();
-  auto currMark = ++mPoints.begin();
+  double h_scale  = mSettings->mGraphics.pHeightScale.get();
```
[the remark this answers is not in the document]
Comment (Schneegans) [reply plugins/csp-measurement-tools/src/PathTool.cpp:217]: True. Not really part of this PR, but a few additional changed lines won't hurt, I guess :smile:

File: plugins/csp-custom-web-ui/src/Plugin.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (JonasGilg) [reply plugins/csp-custom-web-ui/src/Plugin.cpp:133]: It actually doesn't make a difference in this instance. But generally it is best practice to use a universal reference, so you don't forget to use it, when it makes a difference.

File: src/cs-core/Settings.cpp [hunk 11]
```diff
@@ -30,34 +29,109 @@ struct adl_serializer<spdlog::level::level_enum> {
 
 } // namespace nlohmann
 
-namespace cs::core {
+namespace cs::utils {
```
[the remark this answers is not in the document]
Comment (Schneegans) [reply src/cs-core/Settings.cpp:32]: Well, its only the two methods which are required to serialize and deserialize the `cs::utils::ObservableMap<std::string, std::shared_ptr<const cs::scene::CelestialObject>>`. The json library requires that these methods are in the `cs::utils` namespace. However, they are only required in this file and we cannot define them inside of the cs-utils module as they also require classes from cs-scene. So I think it's basically the only viable way here. The only alternative would be to create an `adl_serializer` in the `nlohmann` namespace but I am not sure whether that would be better.

File: src/cs-scene/CelestialAnchor.cpp [hunk 12]
```diff
@@ -72,171 +30,135 @@ CelestialAnchor::CelestialAnchor(std::string sCenterName, std::string sFrameName
...
+
+  if (failed_c()) {
+    std::array<SpiceChar, 320> msg{};
+    getmsg_c("LONG", 320, msg.data());
+    reset_c();
+    throw std::runtime_error(msg.data());
+  }
+
+  // convert to quaternion
+  double axis[3]; // NOLINT(modernize-avoid-c-arrays)
```
[the remark this answers is not in the document]
Comment (Schneegans) [reply src/cs-scene/CelestialAnchor.cpp]: I think there's no apparent reason. I'll fix it.

File: plugins/csp-custom-web-ui/src/Plugin.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (Schneegans) [reply plugins/csp-custom-web-ui/src/Plugin.cpp:133]: I am not sure if this advice should be applied to range-based for-loops. The only "benefit" I could imaging, that your code still compiles and works even if you change the thing you are iterating through to only return const interators. However, I find it very valuable to explicitly differentiate between `auto&` and `auto const&` to make it apparent if I intend to modify the things inside the loop.

What's your view here? Do I miss something?

File: plugins/csp-lod-bodies/src/LodBody.cpp [hunk 13]
```diff
@@ -28,32 +28,20 @@ LodBody::LodBody(std::shared_ptr<cs::core::Settings> const& settings,
...
-  mSettings->initAnchor(*this, anchorName);
-
-  pVisible.connect([this](bool val) {
-    if (val) {
-      mGraphicsEngine->registerCaster(&mPlanet);
-    } else {
-      mGraphicsEngine->unregisterCaster(&mPlanet);
-    }
-  });
+    , mShader(settings, pluginSettings, pGuiManager, mEclipseShadowReceiver) {
```
[the remark this answers is not in the document]
Comment (Schneegans) [reply plugins/csp-lod-bodies/src/LodBody.cpp]: Done.

File: src/cs-utils/ObservableMap.hpp [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+
+  /// Removes the specified element from the container.
+  void erase(K const& key) {
+    auto it = mMap.find(key);
+
+    if (it != mMap.end()) {
+      mOnRemove.emit(key, it->second);
+      mMap.erase(it);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (Schneegans) [reply src/cs-utils/ObservableMap.hpp]: `erase` does not really help here as we still need to emit key and value in with the signal. However there's `extract` which seems to do what we want. I pushed an alternative implementation...

File: src/cs-utils/ObservableMap.hpp [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+
+  /// Inserts a new element into the container if there is no element with the key in the container.
+  void insert(K key, V value) {
+    auto it = mMap.find(key);
+
+    if (it == mMap.end()) {
+      mOnAdd.emit(key, value);
+      mMap.emplace(std::move(key), std::move(value));
+    }
+  }
```
[the remark this answers is not in the document]
Comment (Schneegans) [reply src/cs-utils/ObservableMap.hpp]: To be honest, I do not really understand the advantage of `try_emplace`. I removed the `find` and implemented it with `emplace` now.

File: plugins/csp-custom-web-ui/src/Plugin.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (JonasGilg) [reply plugins/csp-custom-web-ui/src/Plugin.cpp:133]: That is sound reasoning, although the compiler would still throw and error, if you used `auto&&` and tried to modify the value, when is returned as a const.

File: src/cs-utils/ObservableMap.hpp [hunk 16]
```diff
@@ -19,23 +19,22 @@ namespace cs::utils {
...
-  /// This signal will be emitted right before an item is removed from the map.
+  /// This signal will be emitted right after an item was removed from the map.
   Signal<K, V> const& onRemove() const {
     return mOnRemove;
   };
 
   /// Inserts a new element into the container if there is no element with the key in the container.
   void insert(K key, V value) {
-    auto it = mMap.find(key);
+    auto res = mMap.emplace(std::move(key), std::move(value));
```
Comment (JonasGilg) [inline src/cs-utils/ObservableMap.hpp:34]:
```suggestion
    auto [entry, success] = mMap.emplace(std::move(key), std::move(value));
```

File: src/cs-utils/ObservableMap.hpp [hunk 17]
```diff
@@ -19,23 +19,22 @@ namespace cs::utils {
...
 
-    if (it == mMap.end()) {
-      mOnAdd.emit(key, value);
-      mMap.emplace(std::move(key), std::move(value));
+    if (res.second) {
+      mOnAdd.emit(res.first->first, res.first->second);
```
Comment (JonasGilg) [inline src/cs-utils/ObservableMap.hpp:37]:
```suggestion
    if (success) {
      mOnAdd.emit(entry->first, entry->second);
```

**Review by JonasGilg — CHANGES_REQUESTED** [review]
Just some final adjustments.

File: src/cs-scene/CelestialObject.hpp [hunk 18]
```diff
@@ -37,7 +37,7 @@ class CS_SCENE_EXPORT CelestialObject : public CelestialAnchor {
   CelestialObject& operator=(CelestialObject const& other) = default;
   CelestialObject& operator=(CelestialObject&& other) = default;
 
-  virtual ~CelestialObject() = default;
+  virtual ~CelestialObject() override = default;
```
[the remark this answers is not in the document]
Comment (Schneegans) [reply src/cs-scene/CelestialObject.hpp:40]: I think destructors of derived classes are always virtual if the destructor of the base class is virtual. So we can indeed remove it. But is it wrong to leave it there?

[the remark this answers is not in the document]
Comment (JonasGilg) [reply src/cs-scene/CelestialObject.hpp:40]: It is not wrong, but the core guidelines suggest, that only one of `virtual`, `override` or `final` keyword is used:
http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#c128-virtual-functions-should-specify-exactly-one-of-virtual-override-or-final

[the remark this answers is not in the document]
Comment (Schneegans) [reply src/cs-scene/CelestialObject.hpp]: Alright, thanks for the tip!

**Review by JonasGilg — APPROVED** [review]
